File: Java/src/_0045_Jump_Game_II/Solution.java
package _0045_Jump_Game_II;

/**
 * Date: 2021/5/22 21:38
 */
public class Solution {

    public int jump(int[] nums) {
        int n = nums.length;
        int[] f = new int[n];
        f[0] = 0;
        for (int i = 1, j = 0; i < n; i++) {
            while (j + nums[j] < i) j++;
            f[i] = f[j] + 1;
        }
        return f[n - 1];
    }

    public static void main(String[] args) {

        System.out.println((new Solution()).jump(new int[]{2, 3, 1, 1, 4}));  // 2
    }
}

File: Python/_0045_Jump_Game_II/Solution.py
# coding=utf-8
# Date: 2021/5/22 21:40
from typing import List


# 执行用时：40 ms, 在所有 Python3 提交中击败了71.36%的用户
# 内存消耗：14.9 MB, 在所有 Python3 提交中击败了40.73%的用户
class Solution:
    def jump(self, nums: List[int]) -> int:
        n = len(nums)
        f = [0 for _ in range(n)]
        j = 0
        for i in range(1, n):
            while j + nums[j] < i:
                j += 1
            f[i] = f[j] + 1
        return f[n - 1]


if __name__ == "__main__":
    print(Solution().jump([2, 3, 1, 1, 4]))  # 2
